src/core/ext/transport/chttp2/transport/context_list.h

```diff
@@ -0,0 +1,72 @@
+/*
+ *
+ * Copyright 2018 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef GRPC_CORE_EXT_TRANSPORT_CONTEXT_LIST_H
+#define GRPC_CORE_EXT_TRANSPORT_CONTEXT_LIST_H
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/lib/iomgr/buffer_list.h"
+
+#include "src/core/ext/transport/chttp2/transport/internal.h"
+
+namespace grpc_core {
+/** A list of RPC Contexts */
+class ContextList {
+ public:
+  /* Creates a new element with \a context as the value and appends it to the
+   * list. */
+  static void Append(ContextList** head, grpc_chttp2_stream* s) {
+    /* Make sure context is not already present */
+    ContextList* ptr = *head;
+    GRPC_CHTTP2_STREAM_REF(s, "timestamp");
+    while (ptr != nullptr) {
+      if (ptr->s_ == s) {
+        GPR_ASSERT(
+            false &&
+            "Trying to append a stream that is already present in the list");
+      }
+      ptr = ptr->next_;
+    }
+    ContextList* elem = grpc_core::New<ContextList>();
+    elem->s_ = s;
+    if (*head == nullptr) {
+      *head = elem;
+      return;
+    }
+    ptr = *head;
+    while (ptr->next_ != nullptr) {
```

Comment:
Any problem with appending to the front to avoid having to do an O(n) traverse for every Append?

It would need some restructuring. For example, Append could return a pointer to the new head of the list.  And then calling code would do

foo->cl = foo->cl->Append(....);

Other option would be to split up the idea of ContextList and ContextNode and then callers hold onto ContextList